### PR TITLE
Implement timeout feature

### PR DESCRIPTION
**Pull request by rajkumardongre**

### CODE REVIEW

File: src/http/http_client.f90 [hunk 1]
```diff
@@ -109,6 +111,13 @@ function new_request(url, method, header, data, form, file) result(response)
             request%file = file
         end if
 
+        ! Set request timeout.
+        if(present(timeout)) then
+            request%timeout = timeout
+        else
+            request%timeout = -1
+        end if
```
Comment (milancurcic) [inline src/http/http_client.f90]:
```suggestion
        request%timeout = optval(timeout, -1)
```

File: src/http/http_client.f90 [hunk 2]
```diff
@@ -289,6 +301,17 @@ function set_method(curl_ptr, method, response) result(status)
         end select
     end function set_method
 
+    function set_timeout(curl_ptr, timeout) result(status)
+        type(c_ptr), intent(out) :: curl_ptr
+        integer(kind=int64), intent(in) :: timeout
+        integer :: status
```
Comment (milancurcic) [inline src/http/http_client.f90]: Need FORD docstrings here.

**Review by milancurcic — APPROVED** [review]
Thanks @rajkumardongre, mostly looks good, I added a few requests. Also, tests for timeout are missing.

**Comment by rajkumardongre** [comment]
Thanks @milancurcic, I have made the required changes

File: example/timeout.f90 [hunk 3]
```diff
@@ -0,0 +1,18 @@
+program timeout
+    ! This program demonstrates the use of the timeout option. The request below is designed
+    ! to take more than 10 seconds to complete, but we set the timeout value to 5 seconds.
+    ! As a result, the request will fail with an error message that says "Timeout was reached".
+    use http, only: response_type, request
+    implicit none
+    type(response_type) :: response
+
+    ! Delay in response for 10 seconds
+    response = request(url='https://httpbin.org/delay/10', timeout=5)
```
Comment (milancurcic) [inline example/timeout.f90:10]: I suggest making the timeout here shorter to not slow down the test suite too much.

```suggestion
    response = request(url='https://httpbin.org/delay/10', timeout=1)
```

File: src/http/http_client.f90 [hunk 4]
```diff
@@ -289,6 +297,22 @@ function set_method(curl_ptr, method, response) result(status)
...
+        type(c_ptr), intent(out) :: curl_ptr
+            !! Pointer to the curl handle.
+        integer(kind=int64), intent(in) :: timeout
+            !! Timeout seconds for request.
+        integer :: status
+            !! Status code indicating whether the operation was successful.
+        if(timeout < 0) then
+            status = 0
+        else
+            status = curl_easy_setopt(curl_ptr, CURLOPT_TIMEOUT, timeout)
```
Comment (interkosmos) [inline src/http/http_client.f90:327]: Additionally to the read/write timeout `CURLOPT_TIMEOUT`, we can set the connection timeout `CURLOPT_CONNECTTIMEOUT`. The most simple solution is probably to set both to the same timeout value.

**Comment by rajkumardongre** [comment]
Thanks, @interkosmos, I have made the required changes

**Comment by milancurcic** [comment]
@rajkumardongre can you resolve the conflicts (just merging main into this branch may be sufficient) and merge this PR when ready?

**Comment by milancurcic** [comment]
I think httpbin is flaky here (test failing sometimes for basic auth). Any other service we can use for this?

**Comment by rajkumardongre** [comment]
> I think httpbin is flaky here (test failing sometimes for basic auth). Any other service we can use for this?

Thanks @milancurcic, I found https://postman-echo.com/ as an alternative.

File: test/test_auth.f90 [hunk 5]
```diff
@@ -10,7 +10,7 @@ program test_auth
 
     ! setting username and password
     auth = pair_type('user', 'passwd')   
-    res = request(url='https://httpbin.org/basic-auth/user/passwd', auth=auth)
+    res = request(url='https://postman-echo.com/get', auth=auth)
```
Comment (milancurcic) [inline test/test_auth.f90]: Doesn't this need to be?

```suggestion
    res = request(url='https://postman-echo.com/basic-auth', auth=auth)
```

**Comment by rajkumardongre** [comment]
I was considering whether we should make the procedures `pair_has_name`, `get_pair_value`, and `append_pair` from the `http_pair` module available to the top-level API by importing them into the `http` module.

**Comment by milancurcic** [comment]
Is it important for this PR? It seems to me like a separate issue.

We can consider `pair_has_name` (sounds weird, let's re-evaluate the name) and `get_pair_value` if there are good use-cases. `append_pair` I think is a candidate for removal, since Fortran itself has a simpler syntax to do that.

**Comment by rajkumardongre** [comment]
Ok, we can discuss it on other PR

**Review by milancurcic — APPROVED** [review]

**Pull request merged**